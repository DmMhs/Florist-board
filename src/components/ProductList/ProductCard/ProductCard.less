.ProductCard {
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  border: 1px solid #eee;
  box-sizing: border-box;
  border-radius: 8px;
  margin: 4px 2px;
  font-weight: lighter;
  flex-basis: calc(20% - 6px);
  .info {
    cursor: pointer;
    font-size: 1.4rem;
    color: skyblue;
    position: absolute;
    top: 5px;
    right: 5px;
    z-index: 1;
    &:hover {
      color: rgb(0, 183, 255);
    }
  }
  .image {
    width: 165px;
    height: 165px;
    border-radius: 0%;
    border: 1px solid #eee;
    box-shadow: 1px 1px 5px #ccc;
    background-size: cover;
    background-position: center;
    margin-top: 2rem;
    margin-bottom: 1rem;
    .Slider {
      width: 100%;
      height: 100%;
      .LeftArrow {
        left: 0px;
      }
    }
  }
  .title {
    height: 2.5rem;
    font-size: 1.5rem;
    font-style: italic;
  }
  .price {
    height: 2.5rem;
    font-size: 1.2rem;
  }
  .action-panel {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    background-color: whitesmoke;
    width: 100%;
    height: 3rem;
    .action {
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center;
      width: 100%;
      height: 2rem;
      padding: 5px;
    }

    i {
      cursor: pointer;
      font-size: 1.5rem;
      color: #333333;
      &.fa-cart-arrow-down:hover,
      &.active {
        color: rgb(98, 182, 103);
      }
      &.fa-heart:hover,
      &.fa-heart.active-action {
        color: tomato;
      }
    }
  }
}
@media screen and (max-width: 478px) {
  .ProductCard {
    flex-basis: calc(100% - 6px);
  }
}

@media screen and (min-width: 478.1px) and (max-width: 576px) {
  .ProductCard {
    flex-basis: calc(100% - 6px);
  }
}

@media screen and (min-width: 576.1px) and (max-width: 768px) {
  .ProductCard {
    flex-basis: calc(50% - 6px);
  }
}

@media screen and (min-width: 768.1px) and (max-width: 921px) {
  .ProductCard {
    flex-basis: calc(33.3% - 6px);
  }
}

@media screen and (min-width: 921.1px) and (max-width: 1200px) {
  .ProductCard {
    flex-basis: calc(25% - 6px);
    .image {
      width: 140px;
      height: 140px;
    }
  }
}
