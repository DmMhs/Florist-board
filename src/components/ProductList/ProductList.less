.main-wrapper {
  display: flex;
  margin-bottom: 2rem;
}

.cart-toggle {
  display: flex;
  align-items: center;
  position: absolute;
  top: 3rem;
  right: 2rem;
  z-index: 2;
  cursor: pointer;
  font-size: 1.5rem;
  text-decoration: none;
  border-radius: 5px;
  padding: 6px;
  background-color: #eee;
  color: #333333;
  &:hover,
  &:active {
    background-color: tomato;
    color: white;
    transition: 0.3s;
  }
  span {
    margin-left: 10px;
  }
}
.ProductList {
  padding: 2rem;
  margin: 0 auto;
  display: flex;
  flex-wrap: wrap;
  width: 85%;
  transition: 0.5s;
  &.full-width {
    transform: translateX(-7.5%);
    transition: 0.5s;
  }
}

@media screen and (max-width: 478px) {
  .ProductList {
    width: calc(100% - 1.5rem);
  }
  .cart-toggle {
    position: fixed;
    top: auto;
    right: 1.5rem;
    bottom: 1.5rem;
    box-shadow: 1px 1px 5px #ccc;
  }
}

@media screen and (min-width: 478.1px) and (max-width: 576px) {
  .ProductList {
    width: calc(100% - 2rem);
  }
  .cart-toggle {
    position: fixed;
    top: auto;
    right: 2rem;
    bottom: 2rem;
    box-shadow: 2px 2px 5px #ccc;
  }
}

@media screen and (min-width: 576.1px) and (max-width: 768px) {
  .ProductList {
    width: calc(100% - 1rem);
  }
  .cart-toggle {
    position: fixed;
    top: auto;
    right: 1rem;
    bottom: 1rem;
    box-shadow: 2px 2px 5px #ccc;
  }
}

@media screen and (max-width: 920px) {
  .ProductList {
    margin-top: 3rem;
  }
  .cart-toggle {
    position: fixed;
    top: auto;
    right: 1rem;
    bottom: 1rem;
    box-shadow: 2px 2px 5px #ccc;
    font-size: 1.5rem;
    padding: 9px;
  }
}
