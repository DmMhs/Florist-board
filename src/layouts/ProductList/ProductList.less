.main-wrapper {
  display: flex;
  margin-bottom: 2rem;
  width: 100%;
  .popups-wrapper {
    position: fixed;
    bottom: 2rem;
    right: 2rem;
    z-index: 6;
  }
}

.cart-toggle {
  display: flex;
  align-items: center;
  position: absolute;
  top: 3rem;
  right: 2rem;
  z-index: 3;
  cursor: pointer;
  font-size: 1.5rem;
  text-decoration: none;
  border-radius: 5px;
  padding: 6px;
  background-color: #eee;
  color: #333333;
  &:hover,
  &:active {
    background-color: tomato;
    color: white;
    transition: 0.3s;
  }
  span {
    margin-left: 10px;
  }
}
.ProductList {
  margin-left: 2rem;
  padding: 2rem;
  display: flex;
  flex-wrap: wrap;
  width: 80%;
  transition: 0.5s;
  &.full-width {
    transform: translateX(-15%);
    transition: 0.5s;
  }
}

@media screen and (max-width: 478px) {
  .ProductList {
    width: calc(100% - 1.5rem);
  }
  .cart-toggle {
    position: fixed;
    top: auto;
    right: 1.5rem;
    bottom: 1.5rem;
    box-shadow: 1px 1px 5px #ccc;
  }
}

@media screen and (min-width: 478.1px) and (max-width: 576px) {
  .ProductList {
    width: calc(100% - 2rem);
  }
  .cart-toggle {
    position: fixed;
    top: auto;
    right: 2rem;
    bottom: 2rem;
    box-shadow: 2px 2px 5px #ccc;
  }
}

@media screen and (min-width: 576.1px) and (max-width: 768px) {
  .ProductList {
    width: calc(100% - 1rem);
  }
  .cart-toggle {
    position: fixed;
    top: auto;
    right: 1rem;
    bottom: 1rem;
    box-shadow: 2px 2px 5px #ccc;
  }
}

@media screen and (max-width: 920px) {
  .cart-toggle {
    position: fixed;
    top: auto;
    right: 1rem;
    bottom: 1rem;
    box-shadow: 2px 2px 5px #ccc;
    font-size: 1.5rem;
    padding: 9px;
  }
  .main-wrapper {
    display: block;
    width: 100%;
    box-sizing: border-box;
    .popups-wrapper {
      right: 1rem;
      left: 1rem;
      bottom: 1rem;
    }
  }
  .ProductList {
    margin-left: 0;
    width: 0%;
    display: none;
    transition: 0.5s;
    box-sizing: border-box;
    transform: none;
    transition: 0;
    &.full-width {
      width: 100%;
      display: flex;
      transform: none;
      transition: 0;
    }
  }
}
