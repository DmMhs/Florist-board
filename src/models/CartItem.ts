import react from 'react';
export interface CartItem {
  title: string;
  images: string[];
  price: number;
  currency: string;
  available: boolean;
  key: number;
  id: string;
  inCart: boolean;
  description?: string;
  addToCart?:
    | ((event: React.MouseEvent<HTMLElement, MouseEvent>) => void)
    | undefined;
  remove?: void;
  amount?: number;
}
