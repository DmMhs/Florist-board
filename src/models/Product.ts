export interface Product {
  title: string;
  img: string;
  price: number;
  currency: string;
}
