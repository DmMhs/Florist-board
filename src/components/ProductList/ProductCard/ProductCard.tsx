import React, { Component } from 'react';

import './ProductCard.less';

class ProductCard extends Component<
  {
    title: string;
    imgSrc: string;
    price: number;
    currency: string;
  },
  any
> {
  constructor(props: any) {
    super(props);
    this.state = {
      image: 'assets/images/chamomile.jpg',
      title: 'Chamomile',
      price: 10,
      currency: 'usd',
      inCart: false,
      liked: false
    };
  }
  addToCartClickedHandler = (event: any) => {
    if (event.target.classList.contains('active-action')) {
      event.target.classList.remove('active-action');
      this.setState({
        inCart: false
      });
    } else {
      event.target.classList.add('active-action');
      this.setState({
        inCart: true
      });
    }
  };
  likeClickedHandler = (event: any) => {
    if (event.target.classList.contains('active-action')) {
      event.target.classList.remove('active-action');
      this.setState({
        liked: false
      });
    } else {
      event.target.classList.add('active-action');
      this.setState({
        liked: true
      });
    }
  };
  componentDidMount() {
    this.setState({
      image: this.props.imgSrc,
      title: this.props.title,
      price: this.props.price,
      currency: this.props.currency
    });
  }
  render() {
    return (
      <div className="ProductCard">
        <div
          className="image"
          style={{ background: `url(${this.state.image})` }}
        />
        <div className="title">{this.state.title}</div>
        <div className="price">
          {this.state.price}
          {this.state.currency}
        </div>
        <div className="action-panel">
          <i
            className="fas fa-cart-arrow-down"
            onClick={this.addToCartClickedHandler}
          />
          <i className="far fa-heart" onClick={this.likeClickedHandler} />
        </div>
      </div>
    );
  }
}

export default ProductCard;
