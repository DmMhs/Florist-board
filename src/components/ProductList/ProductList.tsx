import React, { Component } from 'react';
import ProductCard from './ProductCard/ProductCard';

import { CartItem } from '../../models/CartItem';
import ShoppingCart from '../ShoppingCart/ShoppingCart';

import './ProductList.less';
import ProductsFilter from './ProductsFilter/ProductsFilter';
import ProductsSort from './ProductsSort/ProductsSort';
import sortByName from '../../services/sortByName';
import sortByPrice from '../../services/sortByPrice';
import Popup from '../Popup/Popup';

interface ProductListProps {
  products: CartItem[];
}
interface ProductListState {
  showCart: boolean;
  cartProducts: CartItem[];
  checkForAvailable: boolean;
  checkForPrice: boolean;
  filterByPrice: {
    from: number | string;
    to: number | string;
  };
  sortBy: string;
  sortOrder: string;
  showFilters: boolean;
  mobileFiltersMode: boolean;
  popupMessages: { id: string; el: JSX.Element }[];
}

let resizeListener: EventListener;

class ProductList extends Component<ProductListProps, ProductListState> {
  private orderByOptionsRef: React.RefObject<HTMLDivElement>;
  private filtersSidebarRef: React.RefObject<HTMLDivElement>;
  private productListRef: React.RefObject<HTMLDivElement>;
  public filterToggleRef: React.RefObject<HTMLDivElement>;
  constructor(props: ProductListProps) {
    super(props);
    this.state = {
      showCart: false,
      cartProducts: [],
      checkForAvailable: false,
      checkForPrice: false,
      filterByPrice: { from: 0, to: Infinity },
      sortBy: 'name',
      sortOrder: 'default',
      showFilters: false,
      mobileFiltersMode: false,
      popupMessages: []
    };
    this.orderByOptionsRef = React.createRef();
    this.filtersSidebarRef = React.createRef();
    this.productListRef = React.createRef();
    this.filterToggleRef = React.createRef();
  }

  componentDidMount() {
    if (window.innerWidth <= 920) {
      this.setState({
        mobileFiltersMode: true
      });
    } else {
      this.setState({
        mobileFiltersMode: false
      });
    }
    resizeListener = () => {
      if (window.innerWidth <= 920) {
        this.setState({
          mobileFiltersMode: true
        });
      } else {
        this.setState({
          mobileFiltersMode: false
        });
      }
    };
    window.addEventListener('resize', resizeListener);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', resizeListener);
  }

  addToCartClickedHandler = (productData: CartItem) => {
    console.log('add to cart');
    if (productData.available) {
      const cartProducts = this.state.cartProducts;
      const index = cartProducts.findIndex((i: CartItem) => {
        return i.id === productData.id;
      });
      const isInCart = index !== -1 ? true : false;
      if (!isInCart) {
        productData.inCart = true;
        this.setState({
          cartProducts: [...cartProducts, productData]
        });
      }
      this.setState(
        {
          popupMessages: [
            ...this.state.popupMessages,
            {
              id: productData.id,
              el: (
                <Popup
                  type="success"
                  message={`Product "${
                    productData.title
                  }" was added to the cart`}
                  key={productData.id}
                />
              )
            }
          ]
        },
        () => {
          const timer = setTimeout(() => {
            const newPopupMessages = [...this.state.popupMessages];
            const index = newPopupMessages.findIndex(
              (i: { id: string; el: JSX.Element }) => {
                return i.id === productData.id;
              }
            );
            newPopupMessages.splice(index, 1);
            this.setState({
              popupMessages: newPopupMessages
            });
          }, 3000);
        }
      );
    }
  };
  toggleCartClickedHandler = (event: React.MouseEvent<HTMLDivElement>) => {
    this.setState({
      showCart: !this.state.showCart
    });
  };
  closeCartClickedHandler = () => {
    this.setState({
      showCart: false
    });
  };
  handleRemoveCartItem = (index: number) => {
    const confirm = window.confirm('Are you sure?');
    if (confirm) {
      this.state.cartProducts[index].inCart = false;
      this.state.cartProducts[index].amount = 1;
      const newProducts = [...this.state.cartProducts];
      newProducts.splice(index, 1);
      this.setState({
        cartProducts: newProducts
      });
    }
  };
  inStockChangedHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({
      checkForAvailable: !this.state.checkForAvailable
    });
  };
  priceFromChangedHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
    const priceFilter = { ...this.state.filterByPrice };
    priceFilter.from = event.target.value === '' ? 0 : +event.target.value;
    this.setState({
      filterByPrice: priceFilter,
      checkForPrice: true
    });
  };

  priceToChangedHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
    const priceFilter = { ...this.state.filterByPrice };
    priceFilter.to = event.target.value === '' ? Infinity : +event.target.value;
    this.setState({
      filterByPrice: priceFilter,
      checkForPrice: true
    });
  };
  sortOrderClickedHandler = (
    event: React.MouseEvent<HTMLDivElement, MouseEvent>
  ) => {
    this.setState({
      sortOrder: this.state.sortOrder === 'inverse' ? 'default' : 'inverse'
    });
  };
  orderByChangedHandler = (
    event: React.MouseEvent<HTMLAnchorElement, MouseEvent>
  ) => {
    this.setState({
      sortBy:
        (event.target as HTMLAnchorElement).innerText === 'name'
          ? 'name'
          : 'price'
    });
  };
  orderByClickedHandler = () => {
    this.orderByOptionsRef.current!.classList.toggle('show');
  };
  filterToggleClickedHandler = () => {
    if (this.filterToggleRef.current === null) {
      this.setState({
        showFilters: !this.state.showFilters
      });
      this.filtersSidebarRef.current!.classList.toggle('hide');
      this.productListRef.current!.classList.toggle('full-width');
    } else {
      this.setState({
        showFilters: !this.state.showFilters
      });
      this.filtersSidebarRef.current!.classList.toggle('hide');
      this.productListRef.current!.classList.toggle('full-width');
    }
  };

  render() {
    const { products } = this.props;
    let productList = products.map((p: CartItem, index: number) => {
      return (
        <ProductCard
          title={p.title}
          images={p.images}
          price={p.price}
          currency={p.currency}
          available={p.available}
          key={index}
          id={p.id}
          inCart={p.inCart}
          addToCart={this.addToCartClickedHandler.bind(this, p as CartItem)}
        />
      );
    });
    const popups = this.state.popupMessages.map(
      (p: { id: string; el: JSX.Element }) => {
        return p.el;
      }
    );
    if (this.state.checkForAvailable === true) {
      productList = productList.filter(item => {
        return item.props.available === true;
      });
    }
    if (this.state.checkForPrice === true) {
      productList = productList.filter(item => {
        return (
          item.props.price >= this.state.filterByPrice.from &&
          item.props.price <= this.state.filterByPrice.to
        );
      });
    }
    if (this.state.sortBy === 'name') {
      productList =
        this.state.sortOrder === 'inverse'
          ? productList.sort(sortByName).reverse()
          : productList.sort(sortByName);
    }
    if (this.state.sortBy === 'price') {
      productList =
        this.state.sortOrder === 'inverse'
          ? productList.sort(sortByPrice).reverse()
          : productList.sort(sortByPrice);
    }
    return (
      <div className="main-wrapper">
        <div className="cart-toggle" onClick={this.toggleCartClickedHandler}>
          <i className="fas fa-shopping-basket" />
          <span>{this.state.cartProducts.length}</span>
        </div>
        <ShoppingCart
          cartItems={this.state.cartProducts}
          showCart={this.state.showCart}
          closeCart={this.closeCartClickedHandler}
          remove={this.handleRemoveCartItem}
        />
        <ProductsFilter
          filterToggle={this.filterToggleClickedHandler}
          filtersSidebarRef={this.filtersSidebarRef}
          filterToggleRef={this.filterToggleRef}
          inStockChanged={this.inStockChangedHandler}
          priceFromChanged={this.priceFromChangedHandler}
          priceToChanged={this.priceToChangedHandler}
          mobileMode={this.state.mobileFiltersMode}
        />
        <div className="ProductList full-width" ref={this.productListRef}>
          <ProductsSort
            sortOrder={this.state.sortOrder}
            sortOrderClicked={this.sortOrderClickedHandler}
            orderByClicked={this.orderByClickedHandler}
            sortBy={this.state.sortBy}
            orderByOptionsRef={this.orderByOptionsRef}
            orderByChanged={this.orderByChangedHandler}
            mobileMode={this.state.mobileFiltersMode}
            filterToggle={this.filterToggleClickedHandler}
          />
          {productList}
        </div>
        <div className="popups-wrapper">{popups}</div>
      </div>
    );
  }
}

export default ProductList;
