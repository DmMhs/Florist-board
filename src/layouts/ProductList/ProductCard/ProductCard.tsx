import React, { Component, RefObject } from 'react';
import { NavLink } from 'react-router-dom';

import { Slider } from '../../../components';
import { CartItem } from '../../../models/CartItem';
import { AppContext } from '../../../AppContext';
import { productsRef } from '../../../firebase';
import './ProductCard.less';
import { deleteProduct } from '../../../services/deleteProduct';

declare global {
  interface Window {
    fbAsyncInit: () => void;
  }
}

interface ProductCardState {
  isLiked: boolean;
  isLikedBy: string[];
}

class ProductCard extends Component<CartItem, ProductCardState> {
  public static getDerivedStateFromProps(
    props: CartItem,
    state: ProductCardState
  ) {
    return {
      isLiked: props.likedBy
    };
  }

  private likeButtonRef: RefObject<HTMLDivElement>;

  constructor(props: CartItem) {
    super(props);
    this.likeButtonRef = React.createRef();
    this.state = {
      isLiked: false,
      isLikedBy: []
    };
  }
  public componentDidMount() {
    if (this.state.isLiked === true) {
      this.likeButtonRef.current!.classList.add('active');
    }
  }

  private likeClickedHandler = async () => {
    this.likeButtonRef.current!.classList.toggle('active');
    await productsRef
      .child(this.props.id)
      .child('likedBy')
      .on('value', snapshot => {
        this.setState({
          isLiked: !this.state.isLiked,
          isLikedBy: snapshot!.val() === null ? [] : snapshot!.val()
        });
      });
    if (this.state.isLiked === true) {
      const newLikedByList = [...this.state.isLikedBy];
      const index = newLikedByList.findIndex((val: string) => {
        return val === this.context.state.userId;
      });
      newLikedByList.splice(index, 1);
      productsRef
        .child(this.props.id)
        .child('likedBy')
        .set(newLikedByList);
    } else {
      await productsRef.child(this.props.id).update({
        likedBy: [...this.state.isLikedBy, this.context.state.userId]
      });
    }
  };

  public render() {
    console.log(this.props);
    const {
      id,
      images,
      title,
      title_ua,
      price,
      currency,
      available,
      inCart
    } = this.props;

    const context = this.context;
    const labels = context.state.labels;
    const lang = context.state.lang;

    const actionIcon = inCart ? (
      <div>
        <div className="action">
          <i className="far fa-check-circle active" />
        </div>
      </div>
    ) : (
      <div>
        <div className="action">
          <i
            className="fas fa-cart-arrow-down"
            onClick={this.props.addToCart}
          />
        </div>
      </div>
    );

    const adminIcons =
      context.state.userRole === 'admin' ? (
        <div className="adminIcons">
          <NavLink to={`/admin/edit-product?id=${id}`}>
            <i className="far fa-edit" />
          </NavLink>
          <i
            className="far fa-trash-alt"
            onClick={deleteProduct.bind(this, id, title.toLowerCase())}
          />
        </div>
      ) : null;

    return (
      <AppContext.Consumer>
        {value =>
          value && (
            <div className="ProductCard">
              {adminIcons}
              <NavLink to={`/product-details/${id}`}>
                <i className="fas fa-info-circle info" />
              </NavLink>
              {value.state.userAuthenticated === true ? (
                <i
                  className="fas fa-heart like"
                  onClick={this.likeClickedHandler}
                  ref={this.likeButtonRef}
                />
              ) : null}
              <div className="image">
                <Slider images={images} auto={false} showControls={true} />
              </div>
              <div className="title">{lang === 'en' ? title : title_ua}</div>
              {available === true ? (
                <div className="price">
                  {price}
                  {currency}
                </div>
              ) : (
                <div className="price">
                  {labels[lang].pages.shop.notAvailable}
                </div>
              )}
              <div className="action-panel">
                {available ? actionIcon : null}{' '}
              </div>
            </div>
          )
        }
      </AppContext.Consumer>
    );
  }
}

ProductCard.contextType = AppContext;

export default ProductCard;
