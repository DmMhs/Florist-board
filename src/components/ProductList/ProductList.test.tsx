import React from 'react';
import { shallow, mount } from 'enzyme';
import ReactDOM from 'react-dom';

import ProductList from './ProductList';
import { Product } from '../../models/Product';
import ProductCard from './ProductCard/ProductCard';
import { CartItem } from '../../models/CartItem';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<ProductList products={[]} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('matches a snapshot', () => {
  const wrapper = shallow(<ProductList products={[]} />);
  expect(wrapper).toMatchSnapshot();
});

it('products list is not empty', () => {
  const prod: Product = {
    title: 'title',
    images: ['pathtoimage'],
    price: 10,
    currency: 'usd',
    available: true
  };
  const wrapper = shallow(<ProductList products={[prod, prod, prod]} />);
  expect(wrapper.find('.ProductList').exists()).toEqual(true);
  expect(wrapper.find('.ProductList').text().length).toBeGreaterThan(0);
});

it('Removes cart items', () => {
  const prod: CartItem = {
    title: 'title',
    images: ['pathtoimage'],
    price: 10,
    currency: 'usd',
    available: true,
    key: 0,
    id: 'safasf',
    inCart: false,
    amount: 1
  };
  const wrapper = shallow(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.state = {
    showCart: false,
    cartProducts: [prod, prod, prod]
  };
  instance.handleRemoveCartItem(2);
  expect(instance.state.cartProducts.length).toEqual(2);
});

it('Add to cart works', () => {
  const prod: CartItem = {
    title: 'title',
    images: ['pathtoimage'],
    price: 10,
    currency: 'usd',
    available: true,
    key: 0,
    id: 'safasf',
    inCart: false,
    amount: 1
  };
  const wrapper = shallow(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.state = {
    showCart: false,
    cartProducts: []
  };
  instance.addToCartClickedHandler(prod);
  expect(instance.state.cartProducts.length).toEqual(1);
});

it('toggleCart works', () => {
  const wrapper = mount(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.state = {
    showCart: false,
    cartProducts: []
  };
  wrapper.find('.cart-toggle').simulate('click');
  expect(instance.state.showCart).toEqual(true);
});

it('state changes on filters toggle click', () => {
  const wrapper = mount(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.state = {
    showFilters: true
  };
  instance.filterToggleClickedHandler();
  expect(instance.state.showFilters).toBeFalsy();
});

it('Filter by price works', () => {
  const wrapper = mount(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.state = {
    showFilters: true,
    checkForPrice: false
  };
  expect(instance.state.checkForPrice).toBeFalsy();
  wrapper.find('.priceFrom').simulate('change');
  expect(instance.state.checkForPrice).toBeTruthy();
  instance.setState({
    showFilters: true,
    checkForPrice: false
  });
  wrapper.find('.priceTo').simulate('change');
  expect(instance.state.checkForPrice).toBeTruthy();
});

it('Click on sort button changes the state', () => {
  const wrapper = mount(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.state = {
    sortBy: 'name'
  };
  wrapper.find('.sort-by-price-btn').simulate('click');
  expect(instance.state.sortBy).toEqual('price');
});

it('sort order button changes the state', () => {
  const wrapper = mount(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.state = {
    sortBy: 'name',
    sortOrder: 'default'
  };
  wrapper.find('.fa-sort-alpha-down').simulate('click');
  expect(instance.state.sortOrder).toEqual('inverse');
});

it('In stock filter changes the state', () => {
  const wrapper = mount(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.state = {
    checkForAvailable: false
  };
  instance.inStockChangedHandler();
  expect(instance.state.checkForAvailable).toBeTruthy();
});

it('Close cart button changes the state', () => {
  const wrapper = mount(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.state = {
    showCart: true
  };
  instance.closeCartClickedHandler();
  expect(instance.state.showCart).toBeFalsy();
});

it('resize event has impact on state', () => {
  const wrapper = shallow(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.setState({
    mobileFiltersMode: undefined
  });
  const resizeListener = () => {
    if (window.innerWidth <= 920) {
      instance.setState({
        mobileFiltersMode: true
      });
      expect(instance.state.mobileFiltersMode).toBeTruthy();
    } else {
      instance.setState({
        mobileFiltersMode: false
      });
      expect(instance.state.mobileFiltersMode).toBeFalsy();
    }
  };
  window.addEventListener('resize', resizeListener);
  let resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
  expect(instance.state.mobileFiltersMode).toBeDefined();
});

fit('filter toggler changes the state', () => {
  const wrapper = mount(<ProductList products={[]} />);
  const instance = wrapper.instance();
  instance.setState({
    showFilters: false
  });
  instance.filterToggleClickedHandler();
  expect(instance.state.showFilters).toBeTruthy();
});
