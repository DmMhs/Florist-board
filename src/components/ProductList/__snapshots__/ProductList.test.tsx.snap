// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches a snapshot 1`] = `
<div>
  <div
    className="cart-toggle"
    onClick={[Function]}
  >
    <i
      className="fas fa-shopping-basket"
    />
    <span>
      0
    </span>
  </div>
  <ShoppingCart
    cartItems={Array []}
    closeCart={[Function]}
    remove={[Function]}
    showCart={false}
  />
  <div
    className="ProductList"
  />
</div>
`;
